[Core] Generic residual based bossak velocity scheme

sunethwarna opened the pull request:
This PR makes a generic residual based bossak velocity scheme, independent of variables. Derived classes of this has to define their own variables in order to use this bossak scheme.

Advantages of this scheme:
1. Can use with any number of different variables, no need to duplicate code.
2. Can specify which variables has second derivatives and other derivative updates.

I've made the initial example in FluidDynamicsApplication residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h

FluidDynamicsApplication's reference values were slightly changed (last few decimal places), therefore those are also updated in the same PR.

sunethwarna committed:
Merge remote-tracking branch 'origin/master' into fluid/validation-test-updates

sunethwarna committed:
adding derivatives extension base class

sunethwarna committed:
adding DERIVATIVES_EXTENSION variable

sunethwarna committed:
adding generic residual based bossak velocity scheme

sunethwarna committed:
adding derrived bossak velocity scheme to fluid dynamics application

sunethwarna committed:
minor changes in ref files

sunethwarna committed:
Merge remote-tracking branch 'origin/master' into generic-residual-based-boassak-velocity-scheme

sunethwarna committed:
minor

philbucher commented:
there are quite some things that can be improved
I will review in the next days, please wait with merging

BTW nice idea! I think this could also be adapted to the schemes for structural!

msandre commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,749 @@
+        mAlphaBossak = AlphaBossak;
+        mUpdateAcceleration = UpdateAcceleration;
+        mUpdateDisplacement = UpdateDisplacement;
+
```
Can you initialize mBossak in some safe/meaningful way? Leaving it like this they will contain garbage values

sunethwarna replied:
done :)

msandre commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,749 @@
+        })");
+        rParameters.ValidateAndAssignDefaults(default_parameters);
+
+        ResidualBasedBossakVelocityScheme(rParameters["alpha_bossak"].GetDouble(),
```
You can't call the constructor like this. You are basically constructing a temporary and throwing it away. Your member data will not be initialized.

sunethwarna replied:
this constructor is removed

msandre commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,749 @@
+        mSecondDerivativeValuesVectorOld.resize(num_threads);
+    }
+
+    ResidualBasedBossakVelocityScheme(Parameters& rParameters)
```
`explicit`

sunethwarna replied:
this constructor is removed

msandre commented on kratos/includes/variables.h:
```diff
@@ -34,6 +34,7 @@
 #include "containers/periodic_variables_container.h"
 #include "utilities/table_stream_utility.h"
 #include "utilities/adjoint_extensions.h"
+#include "utilities/derivatives_extension.h"
```
We should try to think of a way to avoid these includes here (also for adjoint extensions). Basically all of kratos depends on this file.

sunethwarna replied:
I tried using forward declarations, but with kratos::make_unique it gives incomplete type error. Therefore had to leave it like this.

sunethwarna commented:
Thanks for all the comments. I'm closing this PR for the time being. I have  to change the extensions variable from nodal data value container to the elemental data value container. I will reopen it once this is done :)

loumalouomega reviewed (commented):
I see a lot of "repeated code". We should look a way to avoid this...BTW the utility is a good idea, I said some time ago to add the derivatives as a property of the variables itself (or even as a static database)

sunethwarna committed:
updating to element data container

sunethwarna committed:
adding error messages to detect unimplemnted but used methods

sunethwarna committed:
refactor and use elemental data value container

sunethwarna committed:
implement missing methods

sunethwarna committed:
bug fix

sunethwarna commented:
I changed the extension holder to the element data value container, so same scheme can be used with different variables in the same model part by defining different element and their extensions for the respective submodel parts.

Could you please review it again? :)

sunethwarna committed:
Update residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h

philbucher commented on kratos/utilities/derivatives_extension.h:
```diff
@@ -0,0 +1,166 @@
+
+    void load(Serializer& rSerializer)
+    {
+    }
```
I don't think this class should be needing/using the serializer, why did you add it?

jcotela replied:
Anything that can be saved on a variable must be serializable, otherwise you would not be able to serialize nodes/elements.

philbucher replied:
hm ok
It does not even compile if it does not have those methods right?

philbucher commented on kratos/utilities/derivatives_extension.h:
```diff
@@ -0,0 +1,166 @@
+    virtual void GetSecondDerivativesVariables(std::vector<VariableData const*>& rVariables,
+                                               ProcessInfo& rCurrentProcessInfo) const
+    {
+        KRATOS_THROW_ERROR(std::runtime_error,
```
please don't use the old macros any more

sunethwarna replied:
done :+1:

philbucher commented on kratos/utilities/derivatives_extension.h:
```diff
@@ -0,0 +1,166 @@
+ * @ingroup KratosCore
+ * @brief Interface extensions for elements and conditions.
+ */
+class DerivativesExtension
```
this will be really useful from how I understand, thx!

sunethwarna replied:
:+1:

loumalouomega replied:
Can you add Doxygen documentation?

philbucher commented on kratos/utilities/derivatives_extension.h:
```diff
@@ -0,0 +1,166 @@
+#define KRATOS_DERIVATIVES_EXTENSION_INCLUDED
+
+// System includes
+#include <iosfwd>
```
why do you need this?

sunethwarna replied:
removed it

philbucher commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,793 @@
+ * It can be used for either first- or second-order time derivatives. Elements
+ * and conditions must provide a specialization of AdjointExtensions via their
+ * data value container, which allows the scheme to operate independently of
+ * the variable arrangements in the element or condition.
```
copy-paste error

sunethwarna replied:
changed it :+1:

philbucher commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,793 @@
+        const double delta_time = r_current_process_info[DELTA_TIME];
+
+        if (delta_time == 0)
+            KRATOS_THROW_ERROR(
```
old macro and comparing a double with `==`

philbucher commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,793 @@
+
+        BaseType::InitializeSolutionStep(rModelPart, rA, rDx, rb);
+
+        const auto& r_current_process_info = rModelPart.GetProcessInfo();
```
I don't think you need this temporary here

philbucher commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,793 @@
+    void InitializeNodeNeighbourCount(ModelPart::NodesContainerType& rNodes)
+    {
+        // This loop should not be omp parallel
+        // The operation is not threadsafe if the value is uninitialized
```
nope, SetValue is threadsafe
GetValue is not

check my old issues, I discussed this with the others already

btw you should in any case use the variableutils for this

philbucher commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,793 @@
+ * the variable arrangements in the element or condition.
+ */
+template <class TSparseSpace, class TDenseSpace>
+class ResidualBasedBossakVelocityScheme : public Scheme<TSparseSpace, TDenseSpace>
```
some general comments abt this class:
- It seems to be very specific to fluid, which is not a problem, just sad we cannot reuse more code :)
- Please take a look at `utilities/time_discretization` and use the functionalities provided there to avoid copy-pasting of code
- Do you really need the neighbor-elements-stuff? I think this is a pain in MPI and hence difficult to port
- Have you thought how to use this in MPI?

msandre replied:
+1
> Please take a look at utilities/time_discretization and use the functionalities provided there to avoid copy-pasting of code

RiccardoRossi replied:
why specific of the fluid? this is in the core...it should be usable for everryhing

philbucher commented on applications/FluidDynamicsApplication/tests/CylinderTest/cylinder_test_probe1_ref.dat:
```diff
@@ -1,13 +1,13 @@
-0.9 0.5897872427616168 0.04059973181021452 0.45717767292052597
-1 0.588555135752233 0.040614339122423064 0.45119530599827923
-1.1 0.5874297726145277 0.04063544473195884 0.4542584323241518
+0.1 0.6594546997417409 0.03493609518311936 216.24431772568948
```
can you explain why the results change?
From what I understood this PR is refactoring, which should not change the results?

philbucher commented on applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h:
```diff
@@ -77,14 +79,96 @@ namespace Kratos {
+                                             ProcessInfo& rCurrentProcessInfo) const override
+            {
+                rVariables.resize(1);
+                rVariables[0] = &DISPLACEMENT;
```
does this work?
Afaik there are currently some problems with the storage of variables

sunethwarna replied:
This seems to be working :)

philbucher commented on applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h:
```diff
@@ -232,6 +272,25 @@ namespace Kratos {
+            {
+                Element& r_element = *(rModelPart.ElementsBegin() + i);
+                r_element.SetValue(DERIVATIVES_EXTENSION,
+                                        Kratos::make_shared<ElementDerivativesExtension>(&r_element));
```
one per element?
Would it be possible to use one for all? (I don't think so)

Also why shared and not unique ptr?

sunethwarna replied:
thanks :) it is possible to use unique ptr :)

philbucher commented on applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h:
```diff
@@ -281,29 +344,18 @@ namespace Kratos {
                     if (mMeshVelocity == 2)//Lagrangian
                     {
-                        if((itNode)->FastGetSolutionStepValue(IS_LAGRANGIAN_INLET) < 1e-15)
+                        if((itNode)->FastGetSolutionStepValue(IS_LAGRANGIAN_INLET) >= 1e-15)
```
Why changing this?

sunethwarna replied:
Sorry, first i moved this block to a different place, and changed it. When I copied it back, I forgot the order. I changed it back to the original.

philbucher commented on applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h:
```diff
@@ -538,23 +497,7 @@ namespace Kratos {
-            ma4 = pow(DeltaTime, 2)*(-2.0 * mBetaNewmark + 1.0) / 2.0;
-            ma5 = pow(DeltaTime, 2) * mBetaNewmark;
-            mam = (1.0 - mAlphaBossak) / (mGammaNewmark * DeltaTime);
+            BaseType::InitializeSolutionStep(r_model_part, A, Dx, b);
```
better use the `using` syntax if you only call the baseclass

sunethwarna replied:
I don't know the advantage of using "using" syntax over typedef in here. :/ Is there an advantage? @philbucher

philbucher replied:
This function does nothing except calling the baseclass => hence it should be removed
try if it compiles, if not then I can explain you in person what I meant

sunethwarna replied:
Ah, Now I see... It was an empty method calling the base class, which I forgot to remove. I removed it now :) Thanks

philbucher commented on applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h:
```diff
@@ -630,6 +571,12 @@ namespace Kratos {
                 }
             }
+
+            if (mpTurbulenceModel != 0) // If not null
```
```suggestion
            if (mpTurbulenceModel)
```

sunethwarna replied:
changed :)

philbucher reviewed (changes requested):
there are a couple of things that should be improved, but in general a very nice feature that I would also like to adapt to the structural schemes once this PR is merged

sunethwarna committed:
removing old kratos_throw_error and unused header

sunethwarna committed:
replacing if condition

sunethwarna committed:
correct copy paste error

sunethwarna committed:
making a unique_ptr

sunethwarna commented:
@loumalouomega I think I removed and refactored all of the duplicated code. Once you have time, could you look at it? Thanks :)
@philbucher @msandre @jcotela I commited the changes. Could you look at them?

loumalouomega reviewed (dismissed):
OK from my side

sunethwarna committed:
remove empty method

sunethwarna committed:
adding doxygen comments

msandre commented:
@sunethwarna it's in my todo list, but I can't promise a review this week.

sunethwarna committed:
moving the mass matrix term check for efficiency

sunethwarna committed:
making mpDofUpdater protected

sunethwarna committed:
removing unnecessary method

msandre commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,805 @@
+
+        const double delta_time = r_current_process_info[DELTA_TIME];
+
+        if (delta_time == 0)
```
```suggestion
        if (delta_time < std::numeric_limits<double>::epsilon())
```

sunethwarna replied:
updated :)

msandre commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,805 @@
+
+        BaseType::Initialize(rModelPart);
+
+        InitializeNodeNeighbourCount(rModelPart.Nodes());
```
This seems like a very specialized operation that doesn't fit in such a generalized class. Also the fact that it is set but never used by this class is odd.

sunethwarna replied:
removed

msandre commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,805 @@
+            }
+        }
+
+        AssembleVariables(dof_vars, rModelPart.GetCommunicator());
```
Synchronize

sunethwarna replied:
removed

msandre commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,805 @@
+
+    // predicts the solution at the current step as
+    // v = vold
+    void Predict(ModelPart& rModelPart,
```
to be removed (as discussed)

sunethwarna replied:
removed

msandre commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,805 @@
+    ///@name Protected member Variables
+    ///@{
+
+    typename TSparseSpace::DofUpdaterPointerType mpDofUpdater =
```
Making these private would make it easier to change in the future without breaking derived classes.

sunethwarna replied:
done :)

msandre commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,805 @@
+            },
+            r_current_process_info);
+#pragma omp parallel for
+        for (int i = 0; i < number_of_elements; ++i)
```
I think you should try looping over nodes rather than elements here. It is true that the adjoint scheme does it like this, but in the adjoint scheme we need to assemble element contributions in the update step.

sunethwarna replied:
updated as discussed

msandre commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,805 @@
+                r_geometry[i_node].UnSetLock();
+            }
+        }
+        AssembleVariables(acceleration_vars, rModelPart.GetCommunicator());
```
Either you need to synchronize here or do nothing in the case that the variables are already synchronized. This needs to be tested with MPI since serial tests won't show anything and it is a real pain to debug later on.

sunethwarna replied:
updated as discussed

msandre commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,805 @@
+        AssembleVariables(acceleration_vars, rModelPart.GetCommunicator());
+    }
+
+    void UpdateDisplacement(ModelPart& rModelPart)
```
Same comments as for UpdateAcceleration

sunethwarna replied:
updated as discussed

sunethwarna committed:
removied predict method from generic bossak scheme

sunethwarna committed:
moving the dofupdater to privatre scope

sunethwarna committed:
using time_discretization.h

sunethwarna committed:
removing old macros

sunethwarna committed:
changing the variable name to scheme_extension

sunethwarna committed:
removing dofs from the scheme extension interface

RiccardoRossi commented on applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h:
```diff
@@ -80,7 +80,7 @@ namespace Kratos {
     >
     class ResidualBasedPredictorCorrectorVelocityBossakSchemeTurbulent : public ResidualBasedBossakVelocityScheme<TSparseSpace, TDenseSpace> {
-        class ElementDerivativesExtension : public DerivativesExtension
+        class ElementDerivativesExtension : public SchemeExtension
```
i don't know what is being done here ... what is being done here? why do we need to pass such a variable to the element?

sunethwarna replied:
I need the same bossak velocity scheme to be used with different variables. The current implementation had the VELOCITY and PRESSURE variables hard coded in to the scheme, and their derivatives (ACCELERATION) was hard coded as well. With this extension, it allows me to decouple the variables from the scheme and use the same bossak velocity scheme with different variables. This extension provides the variables of the derivatives for the scheme. It needs to be in the element, since I'm using same scheme with different variables in different submodel parts, and I couldn't think of any other way to do it instead of passing it through elements. :/

RiccardoRossi replied:
hi @sunethwarna the point is that doing this comes at a memory price, which you are pushing also for "standard" elements.

as i understood from a quick look into your code, the element is not actually aware of the schemeextension.

at this point why don t you just hold in the scheme a kist of schemeextensions and loop onto that, asking for the element pointer when you need it? this would be actually faster and would have smaller memory overhead than using a pointer variable. also the element woukd not have a variable containing a pointer to the element itself...

RiccardoRossi commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -563,7 +563,7 @@ class ResidualBasedBossakVelocityScheme : public Scheme<TSparseSpace, TDenseSpac
             Element& r_element = *(rModelPart.ElementsBegin() + i);
             Geometry<Node<3>>& r_geometry = r_element.GetGeometry();
-            DerivativesExtension& r_extensions = *r_element.GetValue(DERIVATIVES_EXTENSION);
+            SchemeExtension& r_extensions = *r_element.GetValue(SCHEME_EXTENSION);
```
this definitely does not belong to the base scheme!! we do not want to set a variable for every element.

also as far as i could see it, it was not used in the base scheme

sunethwarna replied:
It is used to get the derivative variables from the element.

RiccardoRossi commented:
what are u trying to accomplish in this PR? i don't like the idea of the scheme adding a variable to each element...

RiccardoRossi commented on applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h:
```diff
@@ -115,24 +186,13 @@ namespace Kratos {
-          Scheme<TSparseSpace, TDenseSpace>(),
+          ResidualBasedBossakVelocityScheme<TSparseSpace, TDenseSpace>(NewAlphaBossak),
           mRotationTool(DomainSize,DomainSize+1,IS_STRUCTURE,0.0), // Second argument is number of matrix rows per node: monolithic elements have velocity and pressure dofs.
           mrPeriodicIdVar(Kratos::Variable<int>::StaticObject())
```
I would argue that periodicity should be handled by MPCs

sunethwarna replied:
@jcotela ? It was there when I ported it to the generic scheme.

RiccardoRossi replied:
yes...consider this comment as a reminder

jcotela replied:
I would leave it on the derived fluid scheme for now.

RiccardoRossi commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,714 @@
+    ///@name Protected member Variables
+    ///@{
+
+    std::vector<LocalSystemVectorType> mSecondDerivativeValuesVectorOld;
```
just to tell that the c++ way would be to use "thread_local" storage (not sure if it works on all compilers though)

RiccardoRossi commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,714 @@
+
+    BossakConstants mBossak;
+
+    std::vector<std::vector<IndirectScalar<double>>> mIndirectCurrentVelocityVector;
```
what is an "IndirectScalar" ? I am lost here

sunethwarna replied:
Sorry, myself @msandre changed the IndirectScalar today. This is used to get the references of derivatives from elements so that they could be updated in the scheme. At the moment, there is no way to pass references of first derivatives and second derivatives from the element. That's why we used this indirect scalar.

RiccardoRossi replied:
could u expand a bit? with a small example maybe...

RiccardoRossi commented on kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h:
```diff
@@ -0,0 +1,714 @@
+        KRATOS_TRY;
+        const int num_threads = OpenMPUtils::GetNumThreads();
+        std::vector<const VariableData*> local_vars;
+        std::vector<std::unordered_set<const VariableData*, Hash, Pred>> thread_vars(num_threads);
```
ouch, here i am really lost. what do you want to accomplish??

aside ot this it is better if your thread variables locally within a parallel region and then you insert them within a critical sectin

~~~cpp
#pragma omp parallel
{
     std::unordered_set<const VariableData*, Hash, Pred> local_var
     ...fill local var...
     #pragma omp critical
     all_vars.insert(local_var.begin(), local_var].end());
}
~~~

sunethwarna replied:
Sorry, this is removed now. I implemented it without having to do the mpi synchronization.

sunethwarna committed:
make indirect scalar based on raw pointers

sunethwarna committed:
using protected update methods with raw variables

sunethwarna committed:
removing the mpi synchronization stuff

RiccardoRossi commented on applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h:
```diff
@@ -360,40 +399,43 @@ namespace Kratos {
-			  itNode->FastGetSolutionStepValue(DISPLACEMENT_X) = 0.0;
-			  itNode->FastGetSolutionStepValue(DISPLACEMENT_Y) = 0.0;
-			}
+                        if((itNode)->FastGetSolutionStepValue(IS_LAGRANGIAN_INLET) < 1e-15)
```
is anyone still using this? maybe we could remove it...

sunethwarna commented:
Hi @RiccardoRossi ,

So indirect scalars are used in two places at the moment. One is at residual_based_adjoint_bossak_scheme.h and as well as in residual_based_bossak_velocity_scheme.h.

In this  indirect scalar, it returns the pointer if it is constructed using the custom constructor, and it returns zero always if it is constructed from the default constructor. Latter behavior is required by the residual_based_adjoint_bossak_scheme.h. It is required because, some of the variables don't have their second derivatives (such as PRESSURE), so it needs to return zero so that inside the scheme, when updating the matrix size need not to be changed.

Indirect scalar holds the pointer to the variable we need to update in the schemes such as ACCELERATION, or any other variable used in the elements.

I am using this generalized bossak scheme for RANS modelling, where at the moment k-epsilon model is developed. In this, what I have done is,
I have a K element, which has TKE as the dof, and TKE_RATE as the time derivative, and an Epsilon element which has Epsilon as the dof and Epsilon_RATE as the time derivative. Element formulation is done in a decoupled way with all the dofs (VELOCITY, PRESSURE, TKE, Epsilon). There are three submodel parts. One for VELOCITY and PRESSURE, second one for TKE element, third one for Epsilon element. Everyone of these elements use the same residual_based_bossak_velocity_scheme.h.

So the scheme asks for elements to get the pointers to their time derivatives which need to be updated after each non-linear iteration. IndirectScalar holds the pointer to the variable, and elements passed SCHEME_EXTENSIONS object which has all the pointers of their time derivatives for their respective nodes. Then all of them is collected by the scheme and then they are updated in the UpdateAcceleration and UpdateDisplacement methods.

Is this the explanation you were expecting?

sunethwarna committed:
removing SCHEME_EXTENSIONS variable from elements

sunethwarna commented:
@RiccardoRossi , hi... I removed the part where I add SCHEME_EXTENSION to the elements. I hope my explanation is clear enough... 

@philbucher I made changed, and removed mpi synchronization stuff which is no longer required. Now it shouldn't have any problems with MPI. The test cases were reverted back to the original reference values since I moved the Predict method from the base class. I think I addressed other comments from you as well. Could please review it? Thank you.

@msandre I made the changes as we discussed. Could you please have a look at it?

All the test cases are passing.

sunethwarna commented:
@philbucher I don't know what is best trade off between execution time and memory consumption, creating it on the fly, or storing it in the elements. Both methods are possible with this scheme. What do you suggest?
As @RiccardoRossi pointed out, storing it on the element is also expensive. Any ideas?

sunethwarna commented:
The other way I can do this is to have three protected methods to get nodal wise update variables for their respective derivatives, which needs to be implemented in the derived class. This will eliminate the complicatations in the UpdateAcceleration and UpdateDisplacement methods w.r.t. OpenMP parallelization. It will remove the dependency on the SchemeExtension as well.

RiccardoRossi reviewed (changes requested):
@sunethwarna i am sorry but i really do not see clearly what you want to do here, or better said how you want to do it.
My big concern is that to my understanding the only item that knows how unknowns and first derivatives are related is the element. for example an element could be ordered internally as

D_X D_Y p D_X D_Y p
D_X D_X  D_Y D_Y p p

or we could have elements that do not have the same variables on all the nodes (think of div stable elements, which for example have a quadratic interpolation of velocity and a linear one for pressure).
This is the reason for which the scheme asks to the elements for the variables it needs, both in terms of first and second variables.
Since this ordering is a design decision on how the elements are implemented, i do not see how you could implement it in a different class.
My problem is that i completely miss understanding how you want to deal with this. Additionally i also miss the role of the IndirectScalar that you are adding as a part of the PR.
Note that now my point is not on the detail of how the IndirectScalar works but on the global picture of how this is intended to be used...

I am sorry but i see this as a reason for blocking the PR until you clarify the design approach you are usig. Note that i am not speaking of coding alone, but rather on the guiding idea behind this

philbucher commented:
@sunethwarna you could take a look at the BDFCustom scheme by @loumalouomega 
https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/solving_strategies/schemes/residual_based_bdf_custom_scheme.h
he does the same as you, but I think it is easier to understand from the outside. Not saying it is better, but I recommend you to take a look

As @RiccardoRossi points out the current approach is quite difficult to understand

msandre commented:
We (@sunethwarna, @RiccardoRossi and @jcotela) discussed this in person. I am summarizing the outcomes of our discussion here (please correct me if I got something wrong):
- instead of storing the extension in the element's data value container, it will be constructed on the fly by the scheme when it is needed through a virtual method (@sunethwarna already pushed this)
- the fluid dynamics scheme will not be modified since there are some concerns about the performance.
- for now, the new schemes will go in the RANS application.

If we agree in the future, we can move some of the features to the core. From my side, the indirect scalar can be refactored as long as it's behavior doesn't change.

sunethwarna commented:
I'm closing this, since I moved this generalized scheme to the https://github.com/KratosMultiphysics/Kratos/tree/RANSModellingApplication-development branch, If the generalization of the scheme is required, we can re open a different PR. :)
